sock: amend API to iterate over stack-internal buffer chunks

miri64 opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
With lwIP we have a chunked UDP payload, so just providing the stack-internal buffer is not possible. To be able to iterate over such a chunked payload, this change allows the `sock_*_recv_buf()` functions to use the internal buffer context as an iteration state. Since the newly required handling of the context requires an iteration anyways, `sock_recv_buf_free()` becomes unnecessary.

As this function is still experimental (I piggy-backed the badge for that) this API change only should require 1 ACK.
<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
Read the doc and

```sh
for test in tests/gnrc_sock_{ip,udp}; do
    make -C ${test} flash test
done
```

should still pass.
<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references
Needed for #13701.
<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

benpicco commented:
> Since GNRC does not chunk the UDP payload, an adaptation for this API change is not required.

That's not entirely true - it must now return 0 when called with `*buf_ctx != NULL`

benpicco commented:
I think the API would be easier to use if instead you introduce an additional `bool *more` parameter.
That way the user doesn't have to call the function twice to find out if there is more data.

miri64 commented:
> I think the API would be easier to use if instead you introduce an additional `bool *more` parameter.
> That way the user doesn't have to call the function twice to find out if there is more data.

I'd like to keep the API minimal. The correct way to use it would be in a while-loop anyway.

benpicco commented:
Just noticed - you should probably also amend `tests/gnrc_sock_udp/main.c`

miri64 commented:
I just had another idea :sweat_smile:

miri64 commented:
Ok, I restructured it a bit and now it somewhat addresses your comment about having to call this function more than once. This is still the case, however, the second call will release the internal context implicitly (what was required to before explicitly using `sock_recv_buf_free()`. So you still need an extra call, but do not require to call `sock_recv_buf_free()` in the end :-).

benpicco commented:
Nice. That 'forces' the user to always use the API properly and recuses the API surface.

I reckon there is no use for the content of `buf_ctx` outside that function?

benpicco commented:
(Murdock has found some issues)

miri64 commented:
> I reckon there is no use for the content of `buf_ctx` outside that function?

Mostly yes. For the lwIP implementation I use the `struct netconn` directly to determine the total length for `-ENOBUFS`, but that stays internal to the lwIP wrapper implementation.

miri64 commented:
> (Murdock has found some issues)

Huh? The change did not run yet.

benpicco commented:
Oh sorry, I confused this with #13701

benpicco commented on sys/net/gnrc/sock/ip/gnrc_sock_ip.c:
```diff
@@ -90,20 +90,21 @@ ssize_t sock_ip_recv(sock_ip_t *sock, void *data, size_t max_len,
         if (res > (ssize_t)max_len) {
-            res = -ENOBUFS;
+            nobufs = true;
+            continue;
```
`break`?
Otherwise this will loop forever.

benpicco replied:
Ah never mind - the next iteration will free `ctx`

miri64 replied:
No, we want to have another call to `sock_ip_recv_buf` to free the buffer space ;-).

benpicco commented:
This looks good now - please squash.

miri64 commented:
> This looks good now - please squash.

Done. Also fixed some doc issues while I was at it ;-).
